feat: Add edit communities

emizzle opened the pull request:
Allow Communities to be edited, including display name, description, color, image, membership, and permissions.

Important changes:
- [x] Something worth noting for reviewers.

Commit:
feat: Add edit communities

Allow Communities to be edited, including display name, description, color, membership, and permissions.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (9)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 8ae8d4ac | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2229/1/) | 2021-05-14 06:57:37 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-210514-065521-8ae8d4.zip) |
| :heavy_check_mark: | 8ae8d4ac | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2229/1/) | 2021-05-14 06:59:33 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-210514-065521-8ae8d4.zip) |
| :heavy_check_mark: | 8ae8d4ac | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2229/1/) | 2021-05-14 07:01:40 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-210514-065521-8ae8d4.aar) |
| | | | | | | |
| :heavy_check_mark: | 118eefe0 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2229/2/) | 2021-05-14 10:07:13 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-210514-100458-118eef.zip) |
| :heavy_check_mark: | 118eefe0 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2229/2/) | 2021-05-14 10:09:02 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-210514-100456-118eef.zip) |
| :heavy_check_mark: | 118eefe0 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2229/2/) | 2021-05-14 10:10:26 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-210514-100455-118eef.aar) |
| | | | | | | |
| :heavy_check_mark: | 378af9d4 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2229/3/) | 2021-05-14 10:16:38 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-210514-101455-378af9.zip) |
| :heavy_check_mark: | 378af9d4 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2229/3/) | 2021-05-14 10:17:25 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-210514-101451-378af9.zip) |
| :heavy_check_mark: | 378af9d4 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2229/3/) | 2021-05-14 10:19:49 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-210514-101451-378af9.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 0a3c729d | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2229/4/) | 2021-05-17 09:13:01 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-210517-091121-0a3c72.zip) |
| :heavy_check_mark: | 0a3c729d | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2229/4/) | 2021-05-17 09:13:07 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-210517-091120-0a3c72.zip) |
| :heavy_check_mark: | 0a3c729d | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2229/4/) | 2021-05-17 09:15:43 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-210517-091118-0a3c72.aar) |
| | | | | | | |
| :heavy_check_mark: | b677a2ef | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2229/5/) | 2021-05-18 11:19:28 | ~9 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-210518-110952-b677a2.aar) |
| :heavy_check_mark: | b677a2ef | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2229/5/) | 2021-05-18 11:20:20 | ~10 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-210518-110952-b677a2.zip) |
| :heavy_check_mark: | b677a2ef | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2229/5/) | 2021-05-18 11:20:37 | ~10 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-210518-110957-b677a2.zip) |

vkjr committed:
Added EditCommunity request type

vkjr committed:
Fix lint errors

vkjr commented:
Few changes:
- added `EditCommunity` request type
- added check that `EditCommunity` invoked by admin
- added publishing community after change

Commit:
Allow editing community without changing image

Previously, retaining an existing community image was not possible because the existing community image path had to be provided in the `editCommunity` RPC call to retain the image. However, once the image is processed by status-go, it is encoded as a base64 string and therefore it is not possible to get the original file path back from this string.

This commit allows for the original to be retained by passing an empty string for the image field in the RPC call.

vkjr commented:
@Samyoul, @cammellos, could you please review?

cammellos commented on protocol/communities/community.go:
```diff
@@ -490,6 +490,14 @@ func (o *Community) BanUserFromCommunity(pk *ecdsa.PublicKey) (*protobuf.Communi
+	o.config.CommunityDescription.Identity.Description = description.Identity.Description
+	o.config.CommunityDescription.Identity.Color = description.Identity.Color
+	o.config.CommunityDescription.Identity.Images = description.Identity.Images
+	o.config.CommunityDescription.Permissions = description.Permissions
```
I would not allow to edit permissions, as that's likely to require some more logic for it to work

cammellos commented on protocol/communities/manager.go:
```diff
@@ -173,13 +183,66 @@ func (m *Manager) CreateCommunity(description *protobuf.CommunityDescription) (*
+		newDescription.Identity.Images = community.config.CommunityDescription.Identity.Images
+	}
+	// TODO: handle delete image (if needed)
+	newDescription.Clock++
```
I don't see clock being copied over/saved in the community, or am I missing something?

vkjr replied:
Great catch, thanks! Yes, we shouldn't adjust clock on description (because it also has incorrect clock). Instead we can call `increaseClock()` right in a community `Edit()` function.

cammellos reviewed (commented):
Looks good mostly, just a couple of questions

vkjr committed:
Don't change permissions. Fixed clock updating

vkjr commented:
@iurimatias, any reason to merge this PR? Related mobile and desktop PRs have the same issue with syncing community changes, so it is not fully working yet.

iurimatias commented:
@vkjr so previously @cammellos had told me that if a PR in status-go had at least one approval I could go ahead and merge it.
Since there was an approval from Andrea himself, and your comments seemed to have been addressed (they say `outdated` and there is commits afterwards), and there were no requests for changes state either, I went ahead and merged it.

iurimatias commented:
@vkjr ah further more, it seems that unfortunatly I merged that right before or around the time that @anastasiyaig posted that issue, unfortunate timing

vkjr commented:
@iurimatias, I see, thanks for explanation!
Probably `status-go` PRs need additional check in PR checklist, something like "Tested on the client"
